src/join/join.rs

```diff
@@ -0,0 +1,374 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+extern crate getopts;
+
+#[macro_use]
+extern crate uucore;
+
+use std::fs::File;
+use std::io::{BufRead, BufReader, Lines, Read, stdin};
+use std::cmp::Ordering;
+
+static NAME: &'static str = "join";
+static VERSION: &'static str = env!("CARGO_PKG_VERSION");
+
+#[derive(PartialEq)]
+enum FileNum {
+    None,
+    File1,
+    File2,
+}
+
+struct Settings {
+    key1: usize,
+    key2: usize,
+    print_unpaired: FileNum,
+    ignore_case: bool,
+}
+
+impl Default for Settings {
+    fn default() -> Settings {
+        Settings {
+            key1: 0,
+            key2: 0,
+            print_unpaired: FileNum::None,
+            ignore_case: false,
+        }
+    }
+}
+
+struct Line {
+    fields: Vec<String>,
+}
+
+impl Line {
+    fn new(string: String) -> Line {
+        Line { fields: string.split_whitespace().map(|s| String::from(s)).collect() }
+    }
+
+    /// Get field at index.
+    fn get_field(&self, index: usize) -> &str {
+        if index < self.fields.len() {
+            &self.fields[index]
+        } else {
+            ""
+        }
+    }
+
+    /// Iterate each field except the one at the index.
+    fn foreach_except<F>(&self, index: usize, f: &F)
+    where
+        F: Fn(&String),
+    {
+        for (i, field) in self.fields.iter().enumerate() {
+            if i != index {
+                f(&field);
+            }
+        }
+    }
+}
+
+struct State {
+    key: usize,
+    print_unpaired: bool,
+    lines: Lines<BufReader<Box<Read>>>,
+    seq: Vec<Line>,
+}
+
+impl State {
+    fn new(name: &str, key: usize, print_unpaired: bool) -> State {
+        let f: Box<Read> = if name == "-" {
+            Box::new(stdin()) as Box<Read>
+        } else {
+            match File::open(name) {
+                Ok(file) => Box::new(file) as Box<Read>,
+                Err(err) => crash!(1, "{}: {}", name, err),
+            }
+        };
+
+        State {
+            key: key,
+            print_unpaired: print_unpaired,
+            lines: BufReader::new(f).lines(),
+            seq: Vec::new(),
+        }
+    }
+
+    /// Compare the key fields of the two current lines.
+    fn compare(&self, other: &State, ignore_case: bool) -> Ordering {
+        let key1 = self.seq[0].get_field(self.key);
+        let key2 = other.seq[0].get_field(other.key);
+
+        compare(key1, key2, ignore_case)
+    }
+
+    /// Skip the current unpaired line.
+    fn skip_line(&mut self) {
+        if self.print_unpaired {
+            self.print_unpaired_line(&self.seq[0]);
+        }
+
+        self.next_line();
+    }
+
+    /// Move to the next line, if any.
+    fn next_line(&mut self) {
+        match self.read_line() {
+            Some(line) => {
+                self.seq[0] = line;
+            }
+            None => {
+                self.seq.clear();
+            }
```

Comment:
In this case (where both branches have only one statement), I would personally not use braces around the branches.